### ISSUE TITLE
Story: Kong Integration

### ISSUE THREAD

**Issue by alexellis** [openfaas/faas#167]
Kong / Mashape have agreed to write a plugin for OpenFaaS in 2018 Q1.

Kong enables interesting features around security and hardening:

* Caching
* rate-limiting
* OAuth tokens
* Basic auth

The OpenFaaS gateway will remain in-tact since its concerns are well defined container management tasks and will receive calls via the Kong API Gateway.

Note: Kong can also be used without a specific OpenFaaS plugin.

**Comment by johnmccabe** [comment]
Derek add label: proposal

**Comment by jonocodes** [comment]
Hi. This is something I might be intersted in. A couple questions though.

Would a plugin really be needed for this? With Kong, you can just point to OpenFaaS as an API and then get access to all the functions provided by that OpenFaaS server. Or presumably just the one in question.

OpenFaaS exposes a REST interface with no credentials out of the box, so its a little simpler then AWS Lambda and OpenWhisk.

**Comment by alexellis** [comment]
I believe you are correct - you can treat the API Gateway as a RESTful interface and use standard Kong features. Would you like to put together a config, template or example?

**Comment by jonocodes** [comment]
Ok. Here is a sample I worked out. You should be able to paste the whole thing into [Play WIth Docker](http://labs.play-with-docker.com/)

Let me know if this is how you or the Kong folk thought it might work.

```
# setup OpenFaaS

docker swarm init --advertise-addr $(hostname -i)

git clone https://github.com/alexellis/faas
cd faas
./deploy_stack.sh
sleep 5

# see that one of the sample functions works

curl localhost:8080/function/func_echoit -d 'hello world'


# setup Kong

docker run -d --name kong-database \
              -p 5432:5432 \
              -e "POSTGRES_USER=kong" \
              -e "POSTGRES_DB=kong" \
              postgres:9.4

docker run --rm \
    --link kong-database:kong-database \
    -e "KONG_DATABASE=postgres" \
    -e "KONG_PG_HOST=kong-database" \
    kong:latest kong migrations up

docker run -d --name kong \
    --link kong-database:kong-database \
    -e "KONG_DATABASE=postgres" \
    -e "KONG_PG_HOST=kong-database" \
    -e "KONG_PROXY_ACCESS_LOG=/dev/stdout" \
    -e "KONG_ADMIN_ACCESS_LOG=/dev/stdout" \
    -e "KONG_PROXY_ERROR_LOG=/dev/stderr" \
    -e "KONG_ADMIN_ERROR_LOG=/dev/stderr" \
    -p 8000:8000 \
    -p 8443:8443 \
    -p 8001:8001 \
    -p 8444:8444 \
    kong:latest
    

# get a route to the OpenFaaS server (surely someone better at Docker swarm networking would not need to do this)

FAASTHOST=`hostname -i`


# Put Kong in front of a single function

sleep 5


curl -i -X POST \
  --url http://localhost:8001/apis/ \
  --data 'name=echoit' \
  --data 'uris=/echo' \
  --data 'upstream_url=http://'$FAASTHOST':8080/function/func_echoit'
  
curl localhost:8000/echo -d 'hello there'

# Or put Kong in front of all the functions

curl -i -X POST \
  --url http://localhost:8001/apis/ \
  --data 'name=functions' \
  --data 'uris=/functs' \
  --data 'upstream_url=http://'$FAASTHOST':8080/function'
  
curl localhost:8000/functs/func_echoit -d 'hello there'


# Add a some auth with a Kong plugin

curl -i -X POST \
  --url http://localhost:8001/apis/echoit/plugins/ \
  --data 'name=key-auth'

curl -i -X POST \
  --url http://localhost:8001/consumers/ \
  --data "username=jdoe"
   
curl -i -X POST \
  --url http://localhost:8001/consumers/jdoe/key-auth/ \
  --data 'key=longsecretkey'


# verify the plugin worked

curl localhost:8000/echo -d 'hello there' --header "apikey: wongkey"
	
curl localhost:8000/echo -d 'hello there' --header "apikey: longsecretkey"
  

```

**Comment by alexellis** [comment]
Thank you for this @jonocodes .. are you already in our Slack community?

Also - I saw Kong can run in-memory, too. Do you have the altered version which doesn't use postgres?

**Comment by ems5311** [comment]
This is awesome because I had the exact same question (whether we need a Kong plugin, or if Kong will just work as-is). If there's no extra work needed to take full advantage of Kong, then that gives OpenFaaS a full API Gateway solution, which we can easily dockerize/automate. Awesome work @jonocodes

**Comment by undernewmanagement** [comment]
@alexellis you mentioned earlier that you read where kong could be used with an in-memory database. I can only find references to cassandra and postgres (i use the latter). Would you happen to remember where you read that?

**Comment by alexellis** [comment]
@undernewmanagement sure it was via the Kong team and it is not available for testing yet. When it is I think we should switch so we can drop the heavy-weight relational DB.

**Comment by coopr** [comment]
Hi, I'm the Product Manager at Mashape / Kong. Currently Kong requires either Postgres or Cassandra. Someday we may support other datastore options.

**Comment by alexellis** [comment]
I'll close this issue since we now have a guide in /guide/ by @jonocodes - I'll also mark #revisit since Mashape have agreed to write a specific integration in Q1 2018.

**Comment by alexellis** [comment]
Derek add label: revisit

**Comment by jonocodes** [comment]
I'm curious what a specific integration would look like since it only takes a single curl to put kong in front of openfaas. Perhaps it will only do that in a Kong plugin? And this would help by giving visibility to openfaas in Kong's plugin list. Besides that would it expose more openfaas functionality?

**Comment by alexellis** [comment]
There are a lot of ad-hoc manual steps required to follow the guide, even then you don't have a "default configuration" that covers all functions - you have to run them in one-by-one right? So an integration into the API would allow for this to be done automatically with a sensible configuration that covers the state as it changes declaratively - not imperatively.

**Comment by jonocodes** [comment]
The single curl here:
https://github.com/openfaas/faas/blob/master/guide/kong_integration.md#use-kong-to-secure-openfaas

is the OpenFaaS equivalent of the Kong Lambda plugin here:
https://getkong.org/plugins/aws-lambda/

All the rest of my doc is boilerplate Kong. I'm just showing how to use their auth plugins.

And in the original readme I provided a way to expose a single function and then another way to expose all the functions. I simplified it to only the latter since that seems like the more common use case. So we are exposing all the functions now. No upkeep needed.
